libvast/test/convertible.cpp

```diff
@@ -0,0 +1,411 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <caf/meta/type_name.hpp>
+
+#include <iterator>
+#define SUITE convertible
+
+#include "vast/concept/convertible/data.hpp"
+#include "vast/concept/parseable/to.hpp"
+#include "vast/concept/parseable/vast/address.hpp"
+#include "vast/concept/parseable/vast/subnet.hpp"
+#include "vast/concept/parseable/vast/time.hpp"
+#include "vast/data.hpp"
+#include "vast/detail/flat_map.hpp"
+#include "vast/test/test.hpp"
+
+#include <caf/test/dsl.hpp>
+
+using namespace vast;
+using namespace vast::test;
+
+template <class From, class To = From>
+struct X {
+  To value;
+
+  template <class Inspector>
+  friend auto inspect(Inspector& fun, X& x) {
+    return fun(x.value);
+  }
+
+  static const record_type layout;
+};
+
+template <class From, class To>
+const record_type X<From, To>::layout = {{"value", data_to_type<From>{}}};
+
+#define BASIC(type, v)                                                         \
```

Comment:
Instead of a macro, what are your thoughts on a function and then the test wrapper calls it for each type? In general, I prefer macros when possible. Something like this:

```c++
auto test_basic = [](auto type, auto v) {
  auto val = type{(v)}; /* NOLINT */                                         
    auto x = X<type>{};                                                        
    auto r = record{{"value", val}};                                           
    REQUIRE_EQUAL(convert(r, x), ec::no_error);                                
    CHECK_EQUAL(x.value, val);
};
```

```c++
TEST(basic-type) {
  test_basic(bool, true);
  test_basic(integer, 42);
  // and so on
}
```

Reply:
In this particular case your suggestion has the drawback that one wouldn't be able to run these tests for specific types with the test name selection for `vast-test` anymore.

In general I think that macros got their poor reputation from a time when people overused or built leaky abstractions with them. Mistakes that the community learned to avoid. Macros can still be used in ways that improve the code when other tools aren't intrusive enough, especially when you don't even have to scroll to find the `#undef`.

Reply:
I have no strong opinion here, but I'm sure @dominiklohmann has. 🙂 

That said, I could see middle-ground where we get the best of both worlds by only wrapping a single line of the `test_basic` piece along with the `TEST` definition.

Reply:
This doesn't look like an improvement to me:
```c++
template <class Type>
auto test_basic = [](auto v) {
  auto val = Type{v};
  auto x = X<Type>{};
  auto r = record{{"value", val}};
  REQUIRE_EQUAL(convert(r, x), ec::no_error);
  CHECK_EQUAL(x.value, val);
};

#define BASIC(type, v)                                                         \
  TEST(basic - type) { /* NOLINT */                                            \
    test_basic<type>(v);                                                       \
  }
```

Reply:
It does to me: now only the TEST definition is macrofied, the rest is "clean" C++.